### PR TITLE
V0.4.16

### PR DESCRIPTION
**Pull request by ViridianMelody**
[Upstream repo](https://github.com/google/jax)

`jaxlib` -> `jax` -> `tensorflow-probability`

If anybody knows more about XLA, I'd like to fix the last line of the test script, but for now I've just commented it out.

### CODE REVIEW

File: recipe/test_jaxlib.py [hunk 1]
```diff
@@ -8,4 +8,10 @@
 y = xops.Sin(x)
 computation = c.Build()
 cpu_backend = xc.make_cpu_client()
-compiled_computation = cpu_backend.compile(computation)
+
```
Comment (cbouss) [inline recipe/test_jaxlib.py:11]: conda-forge has this https://github.com/conda-forge/jaxlib-feedstock/blob/main/recipe/test_jaxlib.py

File: recipe/meta.yaml [hunk 2]
```diff
@@ -7,34 +7,35 @@ package:
...
 
   host:
     - python
     - pip
     - wheel
     - setuptools
-    - numpy 1.21.*  # [py<311]
+    - numpy 1.22.*  # [py<311]
     - numpy 1.23.*  # [py>=311]
+    - python-build
```
[the remark this answers is not in the document]
Comment (JeanChristopheMorinPerso) [reply recipe/meta.yaml:32]: I believe it is required, see https://github.com/google/jax/blob/jaxlib-v0.4.16/jax/tools/build_utils.py#L60.

File: recipe/meta.yaml [hunk 3]
```diff
@@ -7,34 +7,35 @@ package:
...
 
   run:
     - python
-    - numpy >=1.21,<2  # [py<311]
-    - numpy >=1.21,<2  # [py>=311]
-    - scipy >=1.5
+    - numpy >=1.21,<2
+    - scipy >=1.7
+    - ml_dtypes
```
[the remark this answers is not in the document]
Comment (JeanChristopheMorinPerso) [reply recipe/meta.yaml:38]: I'm not sure about that @cbouss , see https://inspector.pypi.io/project/jaxlib/0.4.16/packages/7b/db/c33e8fcdc347a9d437ccd72c8f36bbc9d7bc3334930f35f7f798edeb02b8/jaxlib-0.4.16-cp39-cp39-manylinux2014_x86_64.whl/jaxlib-0.4.16.dist-info/METADATA

```
Requires-Dist: scipy >=1.7
Requires-Dist: numpy >=1.22
Requires-Dist: ml-dtypes >=0.2.0
```

[the remark this answers is not in the document]
Comment (ViridianMelody) [reply recipe/meta.yaml:38]: `opt_einsum` and `importlib-metadata` are dependencies of `jax`, not `jaxlib`. The `jaxlib` dependencies live here: https://github.com/google/jax/blob/jaxlib-v0.4.16/jaxlib/setup.py#L63C12-L63C12

**Comment by skupr-anaconda** [comment]
An error on `osx-arm64`:
```
ERROR: /private/var/tmp/_bazel_builder/4ee1d60ee2fd739fe3f16265a2bf70d1/external/boringssl/BUILD:133:11: Compiling src/crypto/x509v3/v3_alt.c failed: (Exit 1): cc_wrapper.sh failed: error executing command

  (cd /private/var/tmp/_bazel_builder/4ee1d60ee2fd739fe3f16265a2bf70d1/execroot/__main__ && \

  exec env - \
    PATH=/var/folders/k1/30mswbxs7r1g6zwn8y4fyt500000gp/T/abs_daqh1gtob7/croot/jaxlib_1697577680770/_build_env/bin:/var/folders/k1/30mswbxs7r1g6zwn8y4fyt500000gp/T/abs_daqh1gtob7/croot/jaxlib_1697577680770/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placeho/bin:/Users/builder/miniconda3/envs/prefect/condabin:/var/folders/k1/30mswbxs7r1g6zwn8y4fyt500000gp/T/abs_daqh1gtob7/croot/jaxlib_1697577680770/_build_env/bin:/var/folders/k1/30mswbxs7r1g6zwn8y4fyt500000gp/T/abs_daqh1gtob7/croot/jaxlib_1697577680770/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placeho/bin:/Users/builder/miniconda3/envs/prefect/bin:/Users/builder/miniconda3/condabin:/usr/bin:/bin:/usr/sbin:/sbin \

    PWD=/proc/self/cwd \

  external/local_config_cc/cc_wrapper.sh -U_FORTIFY_SOURCE -fstack-protector -Wall -Wthread-safety -Wself-assign -Wunused-but-set-parameter -Wno-free-nonheap-object -fcolor-diagnostics -fno-omit-frame-pointer -g0 -O2 '-D_FORTIFY_SOURCE=1' -DNDEBUG -ffunction-sections -fdata-sections -MD -MF bazel-out/darwin_arm64-opt/bin/external/boringssl/_objs/crypto/v3_alt.pic.d '-frandom-seed=bazel-out/darwin_arm64-opt/bin/external/boringssl/_objs/crypto/v3_alt.pic.o' -fPIC -iquote external/boringssl -iquote bazel-out/darwin_arm64-opt/bin/external/boringssl -isystem external/boringssl/src/include -isystem bazel-out/darwin_arm64-opt/bin/external/boringssl/src/include '-fvisibility=hidden' -Wno-sign-compare '-DMLIR_PYTHON_PACKAGE_PREFIX=jaxlib.mlir.' -DBORINGSSL_IMPLEMENTATION -Wa,--noexecstack -Wall -Werror '-Wformat=2' -Wsign-compare -Wmissing-field-initializers -Wwrite-strings -Wshadow -fno-common '-std=c11' -Wmissing-prototypes -Wold-style-definition -Wstrict-prototypes -no-canonical-prefixes -Wno-builtin-macro-redefined '-D__DATE__="redacted"' '-D__TIMESTAMP__="redacted"' '-D__TIME__="redacted"' -c external/boringssl/src/crypto/x509v3/v3_alt.c -o bazel-out/darwin_arm64-opt/bin/external/boringssl/_objs/crypto/v3_alt.pic.o)

# Configuration: 4a3658130ed594855ec9e148bf40a3576d94bfb3161fa6ed9dee5714f57985c0

# Execution platform: @local_execution_config_platform//:platform

external/boringssl/src/crypto/x509v3/v3_alt.c:57:10: fatal error: 'stdio.h' file not found

#include <stdio.h>
         ^~~~~~~~~

1 error generated.

Target //jaxlib/tools:build_wheel failed to build
```

File: recipe/conda_build_config.yaml [hunk 4]
```diff
@@ -1,6 +1,6 @@
...
-  - 10.14                  # [osx and x86]
-MACOSX_DEPLOYMENT_TARGET:  # [osx and x86]
-  - 10.14                  # [osx and x86]
\ No newline at end of file
+CONDA_BUILD_SYSROOT:       # [osx]
+  - /opt/MacOSX10.14.sdk   # [osx]
+macos_min_version:         # [osx]
+  - 10.14                  # [osx]
+MACOSX_DEPLOYMENT_TARGET:  # [osx]
+  - 10.14                  # [osx]
```
Comment (skupr-anaconda) [inline recipe/conda_build_config.yaml:6]: At the same time osx-arm64 relies on `MacOSX11.x` and not on `10.xx`

**Comment by skupr-anaconda** [comment]
> > An error on osx-arm64:
> 
> This is the same issues I was getting on ray. Basically, it can't find the standard C++ headers provided in the macOS SDK.

I think it's because:
```
macos_min_version:         # [osx]
  - 10.14                  # [osx]
MACOSX_DEPLOYMENT_TARGET:  # [osx]
  - 10.14                  # [osx]
```
But 10.14 should be only for osx-64

**Comment by ViridianMelody** [comment]
> > > An error on osx-arm64:
> > 
> > 
> > This is the same issues I was getting on ray. Basically, it can't find the standard C++ headers provided in the macOS SDK.
> 
> I think it's because:
> 
> ```
> macos_min_version:         # [osx]
>   - 10.14                  # [osx]
> MACOSX_DEPLOYMENT_TARGET:  # [osx]
>   - 10.14                  # [osx]
> ```
> 
> But 10.14 should be only for osx-64

I've been getting this error intermittently on both platforms, both with and without this setting. It builds fine on my local arm64 mac. I'll still make the change, but I suspect this might be an issue specific to `bazel` on Prefect.

**Comment by JeanChristopheMorinPerso** [comment]
>I've been getting this error intermittently on both platforms, both with and without this setting. It builds fine on my local arm64 mac. I'll still make the change, but I suspect this might be an issue specific to bazel on Prefect.

It's partly an issue with bazel and an issue with how the macOS instances are setup. More particularly, the builds will work if run on the old instances (we have 2 osx-64 and 2 osx-arm64). For example, it will fail on my arm64 MBP too.

File: recipe/meta.yaml (hunk 2 above)
[the remark this answers is not in the document]
Comment (cbouss) [reply recipe/meta.yaml:33]: indeed, thank you

**Review by skupr-anaconda — APPROVED** [review]

**Review by JeanChristopheMorinPerso — APPROVED** [review]
Great job @ViridianMelody !

**Pull request merged**